src/common/DecFloat.cpp

```diff
@@ -582,12 +664,34 @@ void Decimal128::toString(string& to) const
 	to.recalculate_length();
 }
 
-double Decimal128::toDouble(DecimalStatus decSt) const
+Decimal128 DecimalFixed::scaled128(DecimalStatus decSt, int scale) const
+{
+	Decimal128 tmp;
+	tmp.set(*this, decSt, -scale);
+	return tmp;
+}
+
+void DecimalFixed::toString(DecimalStatus decSt, int scale, unsigned length, char* to) const
+{
+	scaled128(decSt, scale).toString(decSt, length, to);
+}
+
+void DecimalFixed::toString(DecimalStatus decSt, int scale, string& to) const
+{
+	scaled128(decSt, scale).toString(to);
+}
+
+double Decimal128Base::toDouble(DecimalStatus decSt) const
 {
 	DecimalContext context(this, decSt);
 
 	if (compare(decSt, dmin) < 0 || compare(decSt, dmax) > 0)
 		decContextSetStatus(&context, DEC_Overflow);
+	else if ((!decQuadIsZero(&dec)) && compare(decSt, dzlw) > 0 && compare(decSt, dzup) < 0)
+	{
+		decContextSetStatus(&context, DEC_Underflow);
+		return 0.0;
+	}
 	else
 	{
 		char s[IDecFloat34::STRING_SIZE];
```

Comment:
@AlexPeshkoff here I see:

```
		char s[IDecFloat34::STRING_SIZE];
		memset(s, 0, sizeof(s));
		decQuadToString(&dec, s);
		return atof(s);
```

It seems not necessary to initialize the buffer before call decQuadToString. In utl.cpp you did not it.